### PR TITLE
Fix bug in Flow.scanAsync when upstream completion is delayed

### PR DESCRIPTION
**Pull request by Bobo1239**
I'm not really confident that this is the best solution to the problem (see the new testcase) as I'm new to Akka (or asynchronous programming in general) and it doesn't really "feel" like the proper way but I was unable to think of anything else (which also passed the test).
I'm hoping that someone more knowledgeable can guide me if there's indeed a better solution.
related to: #24036

### CODE REVIEW

**Comment by lightbend-cla-validator** [comment]
Hi @Bobo1239,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by ktoso** [comment]
OK TO TEST

**Comment by Bobo1239** [comment]
Is there anything I can do to move this forward?

**Review by patriknw — COMMENTED** [review]
I think there are more issues

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 1]
```diff
@@ -518,6 +522,8 @@ private[stream] object Collect {
               completeStage()
             case _ ⇒ // in all other cases we will get a complete when the future completes
           }
+        } else if (!futureScheduled && !hasBeenPulled(in)) {
+          completeStage()
         }
```
[the remark this answers is not in the document]
Comment (Bobo1239) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: As I understand it we can't `completeStage` in `onPull` as `onPull` doesn't get called anymore after `onUpstreamFinish` in the supplied test case.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 2]
```diff
@@ -518,6 +522,8 @@ private[stream] object Collect {
               completeStage()
             case _ ⇒ // in all other cases we will get a complete when the future completes
           }
```
[the remark this answers is not in the document]
Comment (Bobo1239) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: That does indeed seem strange. I'll try to find a pathological test case for this and then see what I can do.

**Comment by johanandren** [comment]
Created an issue, closing this now then, thanks for letting us know @Bobo1239

**Pull request closed without merging**